### ISSUE TITLE
Handling of zero or negative degrees-of-freedom

### ISSUE THREAD

**Issue by DougBurke** [sherpa/sherpa#565]
We have a change in behavior of how "unconstrained" fits behave (e.g. those with a degree-of-freedom of 0 or less) compared to the 4.10 behavior, which is related to the replacement of the Fortran-based optimisation code with C++ code (see PR #481). This is the genesis of the tests added in #562. Here is an example of the change (and is related to #562 as the behavior is different to that checked for in the tests). In many ways I think the new behavior is better, since we get an answer rather than an error, with an invalid (`nan`) value for `rstat`, but perhaps we should also set `qval` to `nan` here too.

I note that @dtnguyen2 has just added #564 to address this (as I was writing this issue ;-), but this may be more related to the fit part of the issue, which the tests in #562 don't yet look at [**edited to add** the #562 tests now include fit, and have been partly merged into #564]

With the code

```
from sherpa.data import Data1D
from sherpa.models.basic import Polynom1D
from sherpa.stats import Cash, Chi2, CStat
from sherpa.fit import Fit

d = Data1D('test', [1, 2, 3], [4, 5, 6], [1, 1, 1])
mdl = Polynom1D()
mdl.c1.thaw()
mdl.c2.thaw()
mdl.c0 = 5.1

for stat in [Cash, CStat, Chi2]:
    sobj = stat()
    print("# Stat = {}".format(sobj.name))

    fit = Fit(d, mdl, stat=sobj)
    try:
        statinfo = fit.calc_stat_info()
    except ZeroDivisionError:
        print("\nCAUGHT ZeroDivisionError\n")
        statinfo = None

    print(statinfo)
    print("")
```

WIth CIAO 4.10 I get

```
# Stat = cash
name      = 
ids       = None
bkg_ids   = None
statname  = cash
statval   = -18.2772161919084
numpoints = 3
dof       = 0
qval      = None
rstat     = None

# Stat = cstat

CAUGHT ZeroDivisionError

None

# Stat = chi2

CAUGHT ZeroDivisionError

None

```

With `master` I get

```
# Stat = cash
name      = 
ids       = None
bkg_ids   = None
statname  = cash
statval   = -18.2772161919084
numpoints = 3
dof       = 0
qval      = None
rstat     = None

# Stat = cstat
name      = 
ids       = None
bkg_ids   = None
statname  = cstat
statval   = 0.40863145212838614
numpoints = 3
dof       = 0
qval      = 1.0
rstat     = nan

# Stat = chi2
name      = 
ids       = None
bkg_ids   = None
statname  = chi2
statval   = 2.03
numpoints = 3
dof       = 0
qval      = 1.0
rstat     = nan

```

**Comment by DougBurke** [comment]
Aha - some of this change is intentional, in that Sherpa now (post CIAO 4.10) catches the `ZeroDivisionError` and replaces it with `nan` thanks to https://github.com/sherpa/sherpa/commit/76db8d9194a98f4ac2e141bb8d975840e8312d31#diff-1c90b10b4489451da0fa568740ab4c22 from #486 

Note that there are several places in `sherpa/fit.py` which calculate `rstat` and `qval` and they all take slightly-different approaches. We should consolidate them - e.g. to use the approach of
https://github.com/sherpa/sherpa/blob/master/sherpa/fit.py#L281 - which is 

```
        if _can_calculate_rstat(fit.stat):
            if _dof > 0 and results[2] >= 0.0:
                _qval = igamc(_dof / 2., results[2] / 2.)
                _rstat = results[2] / _dof
            else:
                _rstat = nan
```

although maybe we should also set `_qval` to `nan` rather than leave it as `None` if `_rstat` can not be calculated.

**Comment by DougBurke** [comment]
I have updated the code - see #562 - so that the qval/rstat calculation is now consistent (when returning the fit results and calculating the statistics with `calc_stat_info`), and so that we have a more-consistent behavior for the `rstat` and `qval` fields:

a) both are `None` if the statistic does not support a "goodness of fit"
b) both are `NaN` if they can not be calculated (statistic is < 0 or dof is <= 0); previously qval could be `None` or `1` in this case, depending on the code path
c) otherwise both should be floats (although I guess there's a possibility that `igamc` could return a non-finite value even though the statistic is finite; I would claim this as an "edge case").

**Comment by DougBurke** [comment]
With #562 the script above now prints:

```
# Stat = cash
name      = 
ids       = None
bkg_ids   = None
statname  = cash
statval   = -18.2772161919084
numpoints = 3
dof       = 0
qval      = None
rstat     = None

# Stat = cstat
name      = 
ids       = None
bkg_ids   = None
statname  = cstat
statval   = 0.40863145212838614
numpoints = 3
dof       = 0
qval      = nan
rstat     = nan

# Stat = chi2
name      = 
ids       = None
bkg_ids   = None
statname  = chi2
statval   = 2.03
numpoints = 3
dof       = 0
qval      = nan
rstat     = nan

```

**Comment by DougBurke** [comment]
If we tweak the example script to have -1 dof rather than 0 - that is the following (where I have had to add in an extra catch as the raised error is different from above):

```
from sherpa.data import Data1D
from sherpa.models.basic import Polynom1D
from sherpa.stats import Cash, Chi2, CStat
from sherpa.fit import Fit

d = Data1D('test', [1, 2, 3], [4, 5, 6], [1, 1, 1])
mdl = Polynom1D()
mdl.c1.thaw()
mdl.c2.thaw()
mdl.c3.thaw()
mdl.c0 = 5.1

for stat in [Cash, CStat, Chi2]:
    sobj = stat()
    print("# Stat = {}".format(sobj.name))

    fit = Fit(d, mdl, stat=sobj)
    try:
        statinfo = fit.calc_stat_info()
    except ZeroDivisionError:
        print("\nCAUGHT ZeroDivisionError\n")
        statinfo = None
    except TypeError as exc:
        print("\nCaught {}\n".format(exc))
        statinfo = None

    print(statinfo)
    print("")

```

then with CIAO 4.10 *and* master I get

```
# Stat = cash
name      = 
ids       = None
bkg_ids   = None
statname  = cash
statval   = -18.2772161919084
numpoints = 3
dof       = -1
qval      = None
rstat     = None

# Stat = cstat

Caught igamc domain error, a and x must be positive

None

# Stat = chi2

Caught igamc domain error, a and x must be positive

None

```

with #562 I get

```
# Stat = cash
name      = 
ids       = None
bkg_ids   = None
statname  = cash
statval   = -18.2772161919084
numpoints = 3
dof       = -1
qval      = None
rstat     = None

# Stat = cstat
name      = 
ids       = None
bkg_ids   = None
statname  = cstat
statval   = 0.40863145212838614
numpoints = 3
dof       = -1
qval      = nan
rstat     = nan

# Stat = chi2
name      = 
ids       = None
bkg_ids   = None
statname  = chi2
statval   = 2.03
numpoints = 3
dof       = -1
qval      = nan
rstat     = nan

```

where the differences are

a) we get results from `cstat` and `chi2`
b) you can see this mirrors the `dof=0` case (both `qval` and `rstat` are either `None` or `nan`)

**Comment by DougBurke** [comment]
I've moved the fixes to a new PR, #567, since it was easier than rebasing (due to the way the changes were merged into #562).
